web/src/user_status.ts

```diff
@@ -0,0 +1,144 @@
+import {z} from "zod";
+
+import * as channel from "./channel";
+import * as emoji from "./emoji";
+import type {EmojiRenderingDetails} from "./emoji";
+import {user_settings} from "./user_settings";
+
+export type UserStatus = z.infer<typeof user_status_schema>;
+export type UserStatusEmojiInfo = EmojiRenderingDetails & {
+    emoji_alt_code?: boolean;
+};
+
+type UserStatusEvent = {
+    id: number;
+    type: "user_status";
+    user_id: number;
+    away?: boolean;
+    status_text?: string;
+} & (
+    | {
+          emoji_name: string;
+          emoji_code: string | number;
+          reaction_type: string;
+      }
+    | {
+          emoji_name: undefined;
+      }
+);
+
+const user_status_schema = z.union([
+    z.object({
+        status_text: z.string().optional(),
+        emoji_name: z.string(),
+        emoji_code: z.string().or(z.number()),
+        reaction_type: z.string(),
+        away: z.boolean().optional(),
+    }),
+    z.object({
+        emoji_name: z.undefined(),
+        status_text: z.string().optional(),
+        away: z.boolean().optional(),
+    }),
+]);
+
+const user_status_param_schema = z.record(z.string(), user_status_schema);
+
+const user_info = new Map<number, string>();
+const user_status_emoji_info = new Map<number, UserStatusEmojiInfo>();
+
+export function server_update_status(opts: {
+    status_text?: string;
+    emoji_name?: string;
+    emoji_code?: string | number;
+    reaction_type?: string;
```

Comment:
Under what circumstances are these undefined? I see all the callers passing `|| ""`

Reply:
What was the resolution to this comment?

Reply:
@andersk Tim was right here, I checked all callers of this function and it appears we do infact pass an empty string instead of undefined for these params.